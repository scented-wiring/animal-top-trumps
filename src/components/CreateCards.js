import "../styles/CreateCards.css";
import { useState } from "react";
import { Link } from "react-router-dom";
import axios from "axios";
import Alert from "../components/Alert";
import { addStarterDeck } from "../helpers";

const CreateCards = () => {
  const initialState = {
    fields: {
      name: "",
      aka: "",
      cool: 1,
      largeness: 1,
      handsome: 1,
      alignment: "Lawful Good",
    },
    alert: {
      message: "",
      alertType: "",
    },
  };

  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleAddCard = (event) => {
    event.preventDefault();
    document.getElementsByClassName("createCardsForm")[0].reset();
    if (fields.cool + fields.largeness + fields.handsome > 25) {
      setAlert({
        message:
          "Your card is too powerful! The maximum total for the numerical parameters is 25.",
        alertType: "Error",
      });
    } else {
      axios
        .post("http://localhost:3000/cards", { ...fields })
        .then(() =>
          setAlert({
            message: "Card created.",
            alertType: "Success",
          })
        )
        .catch(() => {
          setAlert({
            message: "Could not connect to the server.",
            alertType: "Error",
          });
        });
    }
  };

  const handleFieldChange = (event) => {
    if (
      event.target.name === "name" ||
      event.target.name === "alignment" ||
      event.target.name === "aka"
    ) {
      setFields({ ...fields, [event.target.name]: event.target.value });
    } else {
      setFields({
        ...fields,
        [event.target.name]: parseInt(event.target.value, 10),
      });
    }
  };

  return (
    <div className="CreateCards">
      <Alert
        setAlert={setAlert}
        message={alert.message}
        alertType={alert.alertType}
      />
      <h2>Card Creator</h2>
      <form className="createCardsForm" onSubmit={handleAddCard}>
        <label htmlFor="name">
          Name
          <input
            id="name"
            name="name"
            autoComplete="off"
            onChange={handleFieldChange}
            maxLength="17"
            pattern="^(?=.*\S).+$"
            required
            title="Must be at least 1 non-space character"
          />
        </label>
        <label htmlFor="cool">
          Cool
          <select
            className="selectNumber"
            name="cool"
            onChange={handleFieldChange}
          >
            <option value="1">1</option>
            <option value="2">2</option>
            <option value="3">3</option>
            <option value="4">4</option>
            <option value="5">5</option>
            <option value="6">6</option>
            <option value="7">7</option>
            <option value="8">8</option>
            <option value="9">9</option>
            <option value="10">10</option>
          </select>
        </label>
        <label htmlFor="largeness">
          Largeness
          <select
            className="selectNumber"
            name="largeness"
            onChange={handleFieldChange}
          >
            <option value="1">1</option>
            <option value="2">2</option>
            <option value="3">3</option>
            <option value="4">4</option>
            <option value="5">5</option>
            <option value="6">6</option>
            <option value="7">7</option>
            <option value="8">8</option>
            <option value="9">9</option>
            <option value="10">10</option>
          </select>
        </label>
        <label htmlFor="handsome">
          Handsome
          <select
            className="selectNumber"
            name="handsome"
            onChange={handleFieldChange}
          >
            <option value="1">1</option>
            <option value="2">2</option>
            <option value="3">3</option>
            <option value="4">4</option>
            <option value="5">5</option>
            <option value="6">6</option>
            <option value="7">7</option>
            <option value="8">8</option>
            <option value="9">9</option>
            <option value="10">10</option>
          </select>
        </label>
        <label htmlFor="aka">
          aka (Nickname)
          <input
            id="aka"
            name="aka"
            autoComplete="off"
            onChange={handleFieldChange}
            maxLength="20"
            pattern="^(?=.*\S).+$"
            required
            title="Must be at least 1 non-space character"
          />
        </label>
        <label htmlFor="alignment">
          Alignment
          <select id="alignment" name="alignment" onChange={handleFieldChange}>
            <option value="Lawful Good">Lawful Good</option>
            <option value="Neutral Good">Neutral Good</option>
            <option value="Chaotic Good">Chaotic Good</option>
            <option value="Lawful Neutral">Lawful Neutral</option>
            <option value="True Neutral">True Neutral</option>
            <option value="Chaotic Neutral">Chaotic Neutral</option>
            <option value="Lawful Evil">Lawful Evil</option>
            <option value="Neutral Evil">Neutral Evil</option>
            <option value="Chaotic Evil">Chaotic Evil</option>
          </select>
        </label>
        <div className="note">
          Note: "aka" and "alignment" are just for fun and do not have any
          bearing on the gameplay.
        </div>
        <button type="submit">Add Card</button>
      </form>
      <div className="starter-deck">
        <div className="starter-text">
          Or press this button to add <br />
          20 starter cards to your deck:
        </div>
        <button type="submit" onClick={() => addStarterDeck(setAlert)}>
          Add Cards
        </button>
      </div>
      <Link to="/">
        <h3>&larr; BACK</h3>
      </Link>
    </div>
  );
};

export default CreateCards;
