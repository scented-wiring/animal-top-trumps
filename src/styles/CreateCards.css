.CreateCards {
  display: flex;
  flex-direction: column;
  align-items: center;
}

h2 {
  font-size: 30px;
  font-family: "Staatliches", sans-serif;
  transition: all 0.4s ease 0s;
}

h2:hover {
  transition: all 0.4s ease 0s;
  color: white;
  letter-spacing: 3px;
}

.createCardsForm {
  width: 250px;
  padding: 20px 30px 20px 30px;
  margin-bottom: 10px;
  display: flex;
  flex-direction: column;
  background-color: #00bdcb;
  border: 3px black solid;
  border-radius: 30px;
}

label {
  display: grid;
  font-size: 20px;
}

.CreateCards select,
input {
  margin-bottom: 10px;
}

.selectNumber {
  width: 100px;
}

.CreateCards button {
  align-self: center;
  height: 30px;
  width: 100px;
  background-color: #73ed8f;
  margin-top: 10px;
  font-family: "Staatliches", sans-serif;
  font-size: 20px;
  border: 2px black solid;
  transition: all 0.4s ease 0s;
}

.CreateCards button:hover {
  background: #385bac;
  color: white;
  width: 120px;
  letter-spacing: 3px;
  transition: all 0.4s ease 0s;
}

.note {
  font-size: 25px;
}
