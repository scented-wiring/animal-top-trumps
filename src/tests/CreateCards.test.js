import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import {
  fireEvent,
  getAllByRole,
  getByRole,
  getByText,
} from "@testing-library/react";
import CreateCards from "../components/CreateCards";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

const renderContainer = () => {
  render(
    <BrowserRouter>
      <CreateCards />
    </BrowserRouter>,
    container
  );
};

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("Create Cards component", () => {
  xtest("renders correctly", () => {
    renderContainer();
    expect(container).toMatchSnapshot();
  });

  xtest("renders home link", () => {
    renderContainer();
    const link = getByRole(container, "link");
    expect(link.textContent).toContain("← BACK");
    expect(link.getAttribute("href")).toBe("/");
  });

  xtest("renders 2 textboxes", () => {
    renderContainer();
    const textboxes = getAllByRole(container, "textbox");
    expect(textboxes).toHaveLength(2);
  });

  xtest("renders 4 select boxes", () => {
    renderContainer();
    const selectBoxes = getAllByRole(container, "combobox");
    expect(selectBoxes).toHaveLength(4);
  });

  xtest("renders a submit button", () => {
    renderContainer();
    const button = getByRole(container, "button");
    expect(button).toBeInTheDocument();
  });

  xtest("renders a success message on submit", async () => {
    var mock = new MockAdapter(axios);
    mock.onPost("http://localhost:3000/cards").reply(200);

    await act(async () => {
      renderContainer();
    });

    await act(async () => {
      fireEvent.click(getByRole(container, "button"));
    });

    const successMessage = getByText(container, "Success!");
    expect(successMessage).toBeInTheDocument();
  });

  xtest("renders an error message on submit", async () => {
    var mock = new MockAdapter(axios);
    mock.onPost("http://localhost:3000/cards").reply(404);

    await act(async () => {
      renderContainer();
    });

    await act(async () => {
      fireEvent.click(getByRole(container, "button"));
    });

    const errorMessage = getByText(container, "Error!");
    expect(errorMessage).toBeInTheDocument();
  });
});
