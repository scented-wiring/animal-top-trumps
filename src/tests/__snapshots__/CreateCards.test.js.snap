// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Cards component renders correctly 1`] = `
<div>
  <div
    class="CreateCards"
  >
    <h2>
      Card Creator
    </h2>
    <form
      class="createCardsForm"
    >
      <label
        for="name"
      >
        Name
        <input
          autocomplete="off"
          id="name"
          maxlength="17"
          name="name"
          pattern="^(?=.*\\\\S).+$"
          required=""
          title="Must be at least 1 non-space character"
        />
      </label>
      <label
        for="cool"
      >
        Cool
        <select
          class="selectNumber"
          name="cool"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </label>
      <label
        for="largeness"
      >
        Largeness
        <select
          class="selectNumber"
          name="largeness"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </label>
      <label
        for="handsome"
      >
        Handsome
        <select
          class="selectNumber"
          name="handsome"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </label>
      <label
        for="aka"
      >
        aka (Nickname)
        <input
          autocomplete="off"
          id="aka"
          maxlength="20"
          name="aka"
          pattern="^(?=.*\\\\S).+$"
          required=""
          title="Must be at least 1 non-space character"
        />
      </label>
      <label
        for="alignment"
      >
        Alignment
        <select
          id="alignment"
          name="alignment"
        >
          <option
            value="Lawful Good"
          >
            Lawful Good
          </option>
          <option
            value="Neutral Good"
          >
            Neutral Good
          </option>
          <option
            value="Chaotic Good"
          >
            Chaotic Good
          </option>
          <option
            value="Lawful Neutral"
          >
            Lawful Neutral
          </option>
          <option
            value="True Neutral"
          >
            True Neutral
          </option>
          <option
            value="Chaotic Neutral"
          >
            Chaotic Neutral
          </option>
          <option
            value="Lawful Evil"
          >
            Lawful Evil
          </option>
          <option
            value="Neutral Evil"
          >
            Neutral Evil
          </option>
          <option
            value="Chaotic Evil"
          >
            Chaotic Evil
          </option>
        </select>
      </label>
      <div
        class="note"
      >
        Note: "aka" and "alignment" are just for fun and do not have any bearing on the gameplay.
      </div>
      <button
        type="submit"
      >
        Add Card
      </button>
    </form>
    <div
      class="starter-deck"
    >
      <div
        class="starter-text"
      >
        Or press this button to add 
        <br />
        20 starter cards to your deck:
      </div>
      <button
        type="submit"
      >
        Add Cards
      </button>
    </div>
    <a
      href="/"
    >
      <h3>
        ‚Üê BACK
      </h3>
    </a>
  </div>
</div>
`;
